tools/rtpsrelay/RelayHandler.cpp

```diff
@@ -907,6 +908,87 @@ SpdpHandler::SpdpHandler(const Config& config,
 : VerticalHandler(config, name, SPDP, address, reactor, guid_partition_table, relay_partition_table, guid_addr_set, rtps_discovery, crypto, application_participant_addr, stats_reporter)
 {}
 
+#ifdef OPENDDS_SECURITY
+namespace {
+  std::string extract_common_name(const OpenDDS::DCPS::Message_Block_Shared_Ptr& msg)
+  {
+    const CORBA::UShort encap_LE = 0x0300; // {PL_CDR_LE} in LE
+    const CORBA::UShort encap_BE = 0x0200; // {PL_CDR_BE} in LE
+
+    OpenDDS::RTPS::MessageParser message_parser(*msg);
+    message_parser.parseHeader();
+
+    while (message_parser.parseSubmessageHeader()) {
+      const auto submessage_header = message_parser.submessageHeader();
+      if (submessage_header.submessageId == OpenDDS::RTPS::DATA) {
+        unsigned short extraFlags;
+        unsigned short octetsToInlineQos;
+        OpenDDS::DCPS::EntityId_t readerId;
+        OpenDDS::DCPS::EntityId_t writerId;
+        OpenDDS::RTPS::SequenceNumber_t writerSequenceNumber;
+
+        if (!(message_parser >> extraFlags) ||
+            !(message_parser >> octetsToInlineQos) ||
+            !(message_parser >> readerId) ||
+            !(message_parser >> writerId) ||
+            !(message_parser >> writerSequenceNumber)) {
+          ACE_ERROR((LM_ERROR, ACE_TEXT("(%P|%t) ERROR: extract_common_name() could not parse submessage\n")));
+          return "";
+        }
+
+        if (writerId != OpenDDS::DCPS::ENTITYID_SPDP_BUILTIN_PARTICIPANT_WRITER) {
+          // Not our message: this could be the same multicast group used
+          // for SEDP and other traffic.
+          break;
+        }
+
+        if (submessage_header.flags & (OpenDDS::RTPS::FLAG_D)) {
```

Comment:
At this point the sequence number has been parsed, so next up is the inlineQos.  If `FLAG_Q` is true, we expect an inlineQos element -- but there's one more wrinkle.  The `octetsToInlineQos` has to be checked (even if no `FLAG_Q`) in case the sender decided to put some additional optional fields in there (a new minor version may add them).

See if other places that parse this have some code that could be borrowed or added to MessageParser.

Reply:
Specifically see `bool operator>>(Serializer& strm,  ::OpenDDS::RTPS::DataSubmessage& stru)` in `RtpsCoreTypeSupportImpl.cpp`

```
  if (!strm.skip(stru.octetsToInlineQos - 16)) {
    return false;
  }
  if ((stru.smHeader.flags & 2) && !(strm >> stru.inlineQos)) {
    return false;
  }
```